src/supervised/ordinal.jl

```diff
@@ -0,0 +1,49 @@
+"""
+    OrdinalMarginLoss <: SupervisedLoss
+
+Modifies a margin loss `loss` to be used on an ordinal domain with
+number of levels `nlevels`. It treats each level as an integer between
+1 and `nlevels`, inclusive, and penalizes output according to the sum of
+level thresholds crossed relative to target
+
+Assumes target is encoded in an Index encoding scheme where levels are
+numbered between 1 and `nlevels`
+"""
+struct OrdinalMarginLoss <: SupervisedLoss
+    loss::MarginLoss
+    nlevels::Int
+end
+
+for fun in (:value, :deriv, :deriv2)
+    @eval @fastmath function ($fun)(loss::OrdinalMarginLoss, target::Number, output::Number)
+        retval = 0
```

Comment:
Sorry, I forgot to provide some more context.

I've redefined 
```julia
struct OrdinalMarginLoss{L<:MarginLoss, N} <: SupervisedLoss
    loss::MarginLoss
end

function OrdinalMarginLoss(loss::T, ::Type{Val{N}}) where {T<:MarginLoss,N}
  typeof(N) <: Number || _serror()
  OrdinalMarginLoss{T,N}(loss)
end
```
As you outlined

I'm trying to define the value, deriv, and deriv2 functions as follows:

```julia
for fun in (:value, :deriv, :deriv2)
    @eval @fastmath @generated function ($fun)(loss::OrdinalMarginLoss{T, N},
                    target::Number, output::Number) where {T <: MarginLoss, N}
        quote
            retval = zero(output)
            @nexprs $N t -> begin
                not_target = (t != target)
                sgn = sign(target - t)
                retval += not_target * ($($fun))(loss.loss, sgn, output - t)
            end
            retval
        end
    end
end
```

It gives the correct answer, but `@code_warntype` still flags retval as being of type ::Any. Is there a different reason for this?